### PR TITLE
Simulate: Add State Change to Exec Trace

### PR DESCRIPTION
**Pull request by ahangsu**
## Summary

This PR implements #5015, which exposes app state (touches) changes in exec trace.

## Test Plan

2 simulation eval test on boxes and app-local/global state.

1 rest client test on app-local/global state.

### CODE REVIEW

File: daemon/algod/api/server/v2/utils.go
Comment (jannotti) [inline daemon/algod/api/server/v2/utils.go:113]: I like the generic version, but I don't know if you need to change the name from `nilToZero`. In Go is common to talk about the "zero value" of a struct, rather than a "default value".

File: data/transactions/logic/opcodeExplain.go
Comment (jannotti) [inline data/transactions/logic/opcodeExplain.go:70]: I wonder if we could get all of these explain functions tested by checking them against what happens in `TestReturnTypes` which is trying every instruction to see if it meets its prototype.

**Review by jannotti — COMMENTED** [review]
Mostly just naming thoughts so far.

File: daemon/algod/api/server/v2/utils.go
Comment (ahangsu) [reply daemon/algod/api/server/v2/utils.go]: right, changed to `nilToZero`.

File: daemon/algod/api/algod.oas2.json
Comment (ahangsu) [reply daemon/algod/api/algod.oas2.json]: changed to `operation` and `application-state`, rather than `application-state-type`.

File: cmd/goal/clerk.go
Comment (ahangsu) [reply cmd/goal/clerk.go]: I renamed to `simulateAppStateChange`.

File: daemon/algod/api/algod.oas2.json
Comment (jannotti) [reply daemon/algod/api/algod.oas2.json]: On app-state, I think the "type" is a useful suffix, because without type, the name "app-state" makes you think the value is going to be actual application state, rather than just which kind of application state is being described.

Also, I'm ok with "app" in json key names, I just want us to say "application" when we explain objects in the "description" fields.

File: daemon/algod/api/algod.oas2.json
Comment (jannotti) [reply daemon/algod/api/algod.oas2.json]: Though I now wonder - do we really need to even report reads in a trace?

Comment (ahangsu) [reply daemon/algod/api/algod.oas2.json]: I am thinking of removing read report from trace, since we want to simply keep track of state change (for now).

File: daemon/algod/api/algod.oas2.json
Comment (ahangsu) [reply daemon/algod/api/algod.oas2.json]: I will change to "g", "l", "b" in response, but see the thoughts above about "crwd"?

File: daemon/algod/api/algod.oas2.json
Comment (jannotti) [reply daemon/algod/api/algod.oas2.json:4111]: Do we need create? If we just report write instead, does a debugger care? Seems like the important thing is to report the new state.

File: daemon/algod/api/algod.oas2.json
Comment (ahangsu) [reply daemon/algod/api/algod.oas2.json]: yea these are definitely separate, so it should be just fine.

File: daemon/algod/api/algod.oas2.json
Comment (ahangsu) [reply daemon/algod/api/algod.oas2.json:4111]: I was thinking about reporting app initial state a bit, where writing to a pre-existing state might be different from creating and writing to a state.

But supposing app initial state contain all the relevant pre-existing states to be touched in simulation, I am perfectly fine with just using write and delete.

File: daemon/algod/api/algod.oas2.json
Comment (ahangsu) [reply daemon/algod/api/algod.oas2.json]: actually, if we don't report read, then we don't have to return app-id, right?

I am taking that out.

File: ledger/simulation/trace.go
Comment (jasonpaulos) [inline ledger/simulation/trace.go:237]: nit: a comment explaining that this will be the zero value for delete operations might be helpful

File: data/transactions/logic/opcodeExplain.go
Comment (jasonpaulos) [inline data/transactions/logic/opcodeExplain.go:27]: nit: this doesn't seem to match our existing naming convention. Why not drop the `T` at the end?

File: data/transactions/logic/opcodeExplain.go
Comment (jasonpaulos) [inline data/transactions/logic/opcodeExplain.go:48]: nit: maybe `BoxState` instead? `logic.Box` seems like too general of a name

File: data/transactions/logic/opcodeExplain.go
Comment (jasonpaulos) [inline data/transactions/logic/opcodeExplain.go:55]: It seems a bit weird that the local state account is passed as a `basics.TealValue`

**Review by jasonpaulos — COMMENTED** [review]
I will take a closer look at testing tomorrow

File: data/transactions/logic/opcodeExplain.go
Comment (ahangsu) [reply data/transactions/logic/opcodeExplain.go]: Renamed to `AppStateOpEnum` to be consistent with `AppStateEnum`.

File: ledger/simulation/trace.go
Comment (ahangsu) [reply ledger/simulation/trace.go:247]: see [jason's pr comment response](https://github.com/algorand/go-algorand/pull/5659/commits/6217847588a6f7ebd07b6490e0be0118e8d7a271)

File: ledger/simulation/trace.go
Comment (ahangsu) [reply ledger/simulation/trace.go]: I thought I can directly take advantage of the name of the struct, without giving a name to the field in struct... fixed in [jason's pr comment response](https://github.com/algorand/go-algorand/pull/5659/commits/6217847588a6f7ebd07b6490e0be0118e8d7a271)

File: daemon/algod/api/algod.oas2.json
Comment (ahangsu) [reply daemon/algod/api/algod.oas2.json:4112]: I exported account field in the object for full quantification of local app state in [jason's pr comment response](https://github.com/algorand/go-algorand/pull/5659/commits/6217847588a6f7ebd07b6490e0be0118e8d7a271)

File: data/transactions/logic/opcodeExplain.go
Comment (ahangsu) [reply data/transactions/logic/opcodeExplain.go]: for now I am returning the stack value representing the local state account directly, do you think it is ok to normalize them into 32-byte representation?

File: ledger/simulation/simulation_eval_test.go
Comment (jasonpaulos) [inline ledger/simulation/simulation_eval_test.go:3207]: nit: use variable `boxContent` here instead?

File: ledger/simulation/simulation_eval_test.go
Comment (jasonpaulos) [inline ledger/simulation/simulation_eval_test.go:3339]: minor: I don't think payment to the app account is required because this test doesn't use boxes?

File: ledger/simulation/simulation_eval_test.go
Comment (jasonpaulos) [inline ledger/simulation/simulation_eval_test.go:3927]: Same here

**Review by jasonpaulos — COMMENTED** [review]
Testing looks good, I just had one real suggestion

File: ledger/simulation/tracer.go
Comment (ahangsu) [reply ledger/simulation/tracer.go:404]: see [new testcase for no new value failure in write to state](https://github.com/algorand/go-algorand/pull/5659/commits/b4cbf3074f2726340e1ed4706b940078d138bc72)

File: ledger/simulation/simulation_eval_test.go
Comment (ahangsu) [reply ledger/simulation/simulation_eval_test.go]: good catch, see [new testcase for no new value failure in write to state](https://github.com/algorand/go-algorand/pull/5659/commits/b4cbf3074f2726340e1ed4706b940078d138bc72)

File: ledger/simulation/simulation_eval_test.go
Comment (ahangsu) [reply ledger/simulation/simulation_eval_test.go]: both are removed in [new testcase for no new value failure in write to state](https://github.com/algorand/go-algorand/pull/5659/commits/b4cbf3074f2726340e1ed4706b940078d138bc72), good catch

**Review by jasonpaulos — DISMISSED** [review]

File: daemon/algod/api/algod.oas2.json
Comment (jannotti) [inline daemon/algod/api/algod.oas2.json:3689]:
```suggestion
          "description": "A boolean option enabling returning application state changes (global, local, and box changes) with the execution trace during simulation.",
```

File: daemon/algod/api/algod.oas2.json
Comment (jannotti) [inline daemon/algod/api/algod.oas2.json:4116]:
```suggestion
          "description": "The key (name) of the global/local/box.",
```

File: daemon/algod/api/algod.oas2.json
Comment (jannotti) [inline daemon/algod/api/algod.oas2.json:4124]:
```suggestion
          "description": "For local state changes, the address of the account associated with the local state.",
```

File: daemon/algod/api/algod.oas3.yml
Comment (jannotti) [inline daemon/algod/api/algod.oas3.yml:1304]:
```suggestion
        "description": "An operation against an application's global/local/box state.",
```

**Review by jannotti — COMMENTED** [review]
Seems good. I made some description language suggestions.

File: daemon/algod/api/server/v2/utils.go
Comment (jannotti) [inline daemon/algod/api/server/v2/utils.go:373]: A global or local key name is allowed to be "" (empty).
Does this do the right thing? I think it might convert to `nil`.

Comment (ahangsu) [reply daemon/algod/api/server/v2/utils.go:373]: hmm, I think you are pointing at wrong place? state key is converted at line 410?

Comment (jannotti) [reply daemon/algod/api/server/v2/utils.go:373]: Yes, totally right. But I suppose it's the same issue. A key's value can be set to "". `sliceOrNil` gives a `nil` when len(s) == 0.

So if someone sets a state value to empty string, we get a nil.

Comment (jannotti) [reply daemon/algod/api/server/v2/utils.go:373]: I guess that's ok? We'll elide the `Bytes` value, but since we said it was of type byte, the debugger can deduce it's the empty string?

Comment (ahangsu) [reply daemon/algod/api/server/v2/utils.go:373]: yes, I was thinking that, the default omitted value get deduced to empty string, and I don't think we allow nil key in states, so this should not be an issue for debugger side.

**Review by jannotti — APPROVED** [review]

**Review by jasonpaulos — DISMISSED** [review]

**Comment by ahangsu** [comment]
Guys, hold off a bit! Wait until #5658 merge in, for we need to extend some testcases from there

**Review by algorandskiy — APPROVED** [review]

**Comment by algorandskiy** [comment]
please re-merge the master

File: data/transactions/logic/opcodeExplain.go
Comment (ahangsu) [reply data/transactions/logic/opcodeExplain.go:64]: It would be a bit involved to modify `TestReturnTypes`, for evaluating generating program give you `cx *EvalContext` **after** evaluating the opcode, while we are unsure what it is like before evaluation. I defer this testing a bit in initial state PR for thorough checks.

**Review by jasonpaulos — APPROVED** [review]

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**
